File: app/src/main/java/com/todoroo/astrid/dao/TaskAttachmentDao.java
/**
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import com.todoroo.astrid.data.RemoteModel;
import com.todoroo.astrid.data.TaskAttachment;
import com.todoroo.astrid.helper.UUIDHelper;

import java.util.List;

@Dao
public abstract class TaskAttachmentDao {

    @Query("SELECT * FROM task_attachments WHERE task_id = :taskUuid AND deleted_at = 0")
    public abstract List<TaskAttachment> getAttachments(String taskUuid);

    @Delete
    public abstract void delete(TaskAttachment taskAttachment);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insert(TaskAttachment attachment);

    @Update
    public abstract void update(TaskAttachment attachment);

    public void createNew(TaskAttachment attachment) {
        if (RemoteModel.isUuidEmpty(attachment.getRemoteId())) {
            attachment.setRemoteId(UUIDHelper.newUUID());
        }
        insert(attachment);
    }
}

File: data/src/commonMain/kotlin/org/tasks/data/dao/TaskAttachmentDao.kt
package org.tasks.data.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import org.tasks.data.entity.Attachment
import org.tasks.data.entity.TaskAttachment

@Dao
interface TaskAttachmentDao {
    @Query("SELECT * FROM attachment WHERE task = :task")
    suspend fun getAttachmentsForTask(task: Long): List<Attachment>

    @Query("SELECT attachment_file.* FROM attachment_file INNER JOIN attachment ON attachment_file.file_uuid = attachment.file_uuid WHERE task = :task")
    suspend fun getAttachments(task: Long): List<TaskAttachment>

    @Query("SELECT * FROM attachment_file")
    suspend fun getAttachments(): List<TaskAttachment>

    @Query("SELECT * FROM attachment_file WHERE file_uuid = :remoteId")
    suspend fun getAttachment(remoteId: String): TaskAttachment?

    @Query("DELETE FROM attachment WHERE task = :taskId AND file_uuid = :attachment")
    suspend fun delete(taskId: Long, attachment: String)

    @Query("DELETE FROM attachment WHERE task = :taskId AND file_uuid IN (:attachments)")
    suspend fun delete(taskId: Long, attachments: List<String>)

    @Insert
    suspend fun insert(attachments: List<Attachment>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(attachment: TaskAttachment)

    @Update
    suspend fun update(attachment: TaskAttachment)

    @Delete
    suspend fun delete(value: List<TaskAttachment>)
}